File: FrontEnd/src/algorithms/simulated-annealing-algorithm.ts
import { BaseAgent } from "./base-agent";
import { normalize } from "@/utils/game-utils";

/**
 * Simulated Annealing Algorithm
 * Probabilistic acceptance of worse solutions to escape local optima
 */
export class SimulatedAnnealingAgent extends BaseAgent {
  constructor(wordList: string[]) {
    super(wordList.map(w => normalize(w)));
  }

  protected chooseBestGuess(possibleWords: string[]): string {
    // If only one word left, return it
    if (possibleWords.length === 1) {
      return possibleWords[0];
    }

    // First guess: use optimal starter
    if (this.currentGuesses.length === 0) {
      const starters = this.getOptimalStarters();
      for (const starter of starters) {
        if (possibleWords.includes(starter)) {
          return starter;
        }
      }
    }

    // Simulated Annealing parameters
    let temperature = 100.0;
    const coolingRate = 0.85;
    const minTemperature = 1.0;
    const maxIterations = 50;

    // Start from random word
    let currentWord = possibleWords[Math.floor(Math.random() * Math.min(10, possibleWords.length))];
    let currentScore = this.scoreWordEntropy(currentWord, possibleWords);
    let bestWord = currentWord;
    let bestScore = currentScore;

    let iterations = 0;

    while (temperature > minTemperature && iterations < maxIterations) {
      iterations++;

      // Get random neighbor
      const neighbors = this.getNeighbors(currentWord, possibleWords, 20);
      if (neighbors.length === 0) break;

      const neighbor = neighbors[Math.floor(Math.random() * neighbors.length)];
      const neighborScore = this.scoreWordEntropy(neighbor, possibleWords);

      // Calculate acceptance probability
      const delta = neighborScore - currentScore;
      const acceptanceProbability = delta > 0 ? 1 : Math.exp(delta / temperature);

      // Accept or reject based on probability
      if (Math.random() < acceptanceProbability) {
        currentWord = neighbor;
        currentScore = neighborScore;

        // Update best if improved
        if (currentScore > bestScore) {
          bestWord = currentWord;
          bestScore = currentScore;
        }
      }

      // Cool down temperature
      temperature *= coolingRate;
    }

    return bestWord;
  }

  /**
   * Get neighbors of a word (random sample from possible words)
   */
  private getNeighbors(word: string, possibleWords: string[], sampleSize: number): string[] {
    const neighbors: string[] = [];
    
    // Add random samples from possible words
    const shuffled = [...possibleWords].sort(() => Math.random() - 0.5);
    neighbors.push(...shuffled.slice(0, sampleSize));

    // Remove duplicates and the current word
    return Array.from(new Set(neighbors)).filter(w => w !== word);
  }
}

File: Backend/algorithms/simulated_annealing_algorithm.py
"""
Simulated Annealing Algorithm

Probabilistic algorithm that can escape local optima by accepting 
worse solutions with decreasing probability over time.
"""

from typing import List
import random
import math
from .base_agent import BaseAgent


class SimulatedAnnealingAgent(BaseAgent):
    """
    Simulated Annealing Algorithm Agent
    
    Strategy: Probabilistic search with temperature cooling
    - Starts from random word
    - Accepts better solutions always
    - Accepts worse solutions with probability e^(Δ/T)
    - Temperature decreases over time (cooling)
    - Can escape local optima
    """
    
    def choose_best_guess(self, possible_words: List[str]) -> str:
        """
        Choose the best guess using simulated annealing
        
        Args:
            possible_words: List of words that match all previous patterns
            
        Returns:
            The best word found by simulated annealing
        """
        # If only one word left, return it
        if len(possible_words) == 1:
            return possible_words[0]
        
        # First guess: use optimal starter
        if len(self.current_guesses) == 0:
            starters = self.get_optimal_starters()
            for starter in starters:
                if starter in possible_words:
                    return starter
        
        # Simulated Annealing parameters
        temperature = 100.0
        cooling_rate = 0.85
        min_temperature = 1.0
        max_iterations = 50
        
        # Start from random word
        start_index = random.randint(0, min(9, len(possible_words) - 1))
        current_word = possible_words[start_index]
        current_score = self.score_word_entropy(current_word, possible_words)
        best_word = current_word
        best_score = current_score
        
        iterations = 0
        
        while temperature > min_temperature and iterations < max_iterations:
            iterations += 1
            
            # Get random neighbor
            neighbors = self._get_neighbors(current_word, possible_words, 20)
            if not neighbors:
                break
            
            neighbor = random.choice(neighbors)
            neighbor_score = self.score_word_entropy(neighbor, possible_words)
            
            # Calculate acceptance probability
            delta = neighbor_score - current_score
            if delta > 0:
                acceptance_probability = 1.0
            else:
                acceptance_probability = math.exp(delta / temperature)
            
            # Accept or reject based on probability
            if random.random() < acceptance_probability:
                current_word = neighbor
                current_score = neighbor_score
                
                # Update best if improved
                if current_score > best_score:
                    best_word = current_word
                    best_score = current_score
            
            # Cool down temperature
            temperature *= cooling_rate
        
        return best_word
    
    def _get_neighbors(self, word: str, possible_words: List[str], sample_size: int) -> List[str]:
        """
        Get neighbors of a word (random sample from possible words)
        
        Args:
            word: Current word
            possible_words: List of all possible words
            sample_size: Number of neighbors to sample
            
        Returns:
            List of neighbor words
        """
        neighbors = []
        
        # Random sample from possible words
        shuffled = possible_words.copy()
        random.shuffle(shuffled)
        neighbors.extend(shuffled[:sample_size])
        
        # Remove duplicates and the current word
        neighbors = list(set(neighbors))
        if word in neighbors:
            neighbors.remove(word)
        
        return neighbors
